python/gym_ignition/randomizers/model/sdf.py

```diff
@@ -0,0 +1,235 @@
+        return self._root.findall(xpath)
+
+    def process_data(self) -> None:
+
+        # Since we support multi-match XPaths, we expand all the individual matches
+        expanded_randomizations = []
+
+        for data in self._randomizations:
+
+            # Find all the matches
+            elements: List[etree.Element] = self._root.findall(path=data.xpath)
+
+            if len(elements) == 0:
+                raise RuntimeError(f"Failed to find elements from XPath '{data.xpath}'")
+
+            for element in elements:
+
+                if data.ignore_zeros and float(self._get_element_text(element)) == 0:
+                    continue
+
+                # Get the precise XPath to the element
+                element_xpath = element.getroottree().getpath(element)
+
+                # Get the parameters
+                params = data.parameters
+
+                if data.method in {Method.Additive, Method.Coefficient}:
+                    element_text = float(self._get_element_text(element))
+                    self._default_values[element] = element_text
+
+                # Update the data
+                complete_data = data._replace(
+                    xpath=element_xpath, element=element, parameters=params)
+
+                expanded_randomizations.append(complete_data)
+
+        # Store the updated data
+        self._randomizations = expanded_randomizations
+
+    def sample(self, pretty_print=False) -> str:
+
+        for data in self._randomizations:
+
+            if data.distribution is Distribution.Gaussian:
+
+                sample = self.rng.normal(loc=data.parameters.mean,
+                                         scale=data.parameters.variance)
+
+            elif data.distribution is Distribution.Uniform:
+
+                sample = self.rng.uniform(low=data.parameters.low,
+                                          high=data.parameters.high)
+
+            else:
+                raise ValueError("Distribution not recognized")
+
+            if data.force_positive:
+                sample = max(sample, 0.0)
+
+            # Update the value
+            if data.method is Method.Absolute:
+
+                data.element.text = str(sample)
+
+            elif data.method is Method.Additive:
+
+                default_value = self._default_values[data.element]
+                data.element.text = str(sample + default_value)
+
+            elif data.method is Method.Coefficient:
+
+                default_value = self._default_values[data.element]
+                data.element.text = str(sample * default_value)
+
+            else:
+                raise ValueError("Method not recognized")
+
+        return etree.tostring(self._root, pretty_print=pretty_print).decode()
+
+    def new_randomization(self) -> RandomizationDataBuilder:
+        return RandomizationDataBuilder(randomizer=self)
+
+    def insert(self, randomization_data) -> None:
+        self._randomizations.append(randomization_data)
```

Comment:
Why are the randomization configuration data appended? How is `self._randomization` organized? Is it a list with a `randomization_data` configuration object for each element to be randomized?

Reply:
Tip: always check the type hinting, they're there to help you understand the type of the variables even in a language that is not statically typed.

You will find that `_randomizations` is a `List[RandomizationData]`. It is used in the `add` method of the `*Builder` to insert the new randomization in a buffer of this class.

Reply:
See documentation ed0c1cb.

Reply:
Ok, thanks.